app/services/hyrax/file_set_characterizer.rb

```diff
@@ -0,0 +1,55 @@
+# frozen_string_literal: true
+
+module Hyrax
+  ##
+  # Characterizes an ActiveFedora based FileSet
+  class FileSetCharacterizer
+    ##
+    # @!attribute [rw] source
+    #   @return [#characterize]
+    attr_accessor :source
+
+    ##
+    # @param source the object to characterize
+    def initialize(source:)
+      self.source = source
+    end
+
+    ##
+    # @return [void]
+    #
+    # @raise [RuntimeError] if FileSet is missing the characterization_proxy
+    def characterize
+      Hydra::Works::CharacterizationService.run(characterization_proxy, filepath)
+      Rails.logger.debug "Ran characterization on #{characterization_proxy.id} (#{characterization_proxy.mime_type})"
+      characterization_proxy.alpha_channels = channels(filepath) if source.image? && Hyrax.config.iiif_image_server?
+      characterization_proxy.save!
+      update_source
+      CreateDerivativesJob.perform_later(source, source.original_file.id, filepath)
+    end
+
+    private
+
+      def update_source
+        source.update_index
+        source.parent&.in_collections&.each(&:update_index)
+      end
+
+      def characterization_proxy
+        raise "#{source.class.characterization_proxy} was not found for FileSet #{source.id}" unless source.characterization_proxy?
+        source.characterization_proxy
+      end
```

Comment:
This method is called five times, I think, which means a bunch of unnecessary checks to see if the proxy exists. Would you be open to inlining this back up in `#characterize`?